code/game/objects/structures/safe.dm

```diff
@@ -190,6 +96,141 @@ FLOOR SAFES
 			if(3)
 				desc = initial(desc) + "\nThe lock seems to be broken."
 
+/obj/structure/safe/ui_assets(mob/user)
+	return list(
+		get_asset_datum(/datum/asset/simple/safe),
+	)
+
+/obj/structure/safe/ui_state(mob/user)
+	return GLOB.physical_state
+
+/obj/structure/safe/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "Safe", name)
+		ui.open()
+		ui.set_autoupdate(FALSE)
+
+/obj/structure/safe/ui_data(mob/user)
+	var/list/data = list()
+	data["dial"] = dial
+	data["open"] = open
+	data["locked"] = locked
+	data["broken"] = broken
+	if(explosion_count > BROKEN_THRESHOLD)
+		data["broken"] = TRUE
+
+	if(open)
+		var/list/contents_names = list()
+		data["contents"] = contents_names
+		for(var/obj/O in contents)
+			contents_names[++contents_names.len] = list("name" = O.name, "sprite" = O.icon_state)
+			user << browse_rsc(icon(O.icon, O.icon_state), "[O.icon_state].png")
```

Comment:
Let's hope MSO doesn't see this.

Reply:
I have no problem with taking it out if needed, its just icons for items inside the safe, which looks quite neat.

Reply:
Yeah it does.